horovod/run/elastic/driver.py

```diff
@@ -0,0 +1,283 @@
+        return self._host_assignments[host][slot] if self.has_rank_assignment(host, slot) \
+            else hosts.INVALID_SLOT_INFO
+
+    def has_rank_assignment(self, host, slot):
+        if self._host_manager.is_blacklisted(host):
+            return False
+        return host in self._host_assignments and len(self._host_assignments[host]) > slot
+
+    def wait_for_available_hosts(self, min_np, min_hosts=1):
+        extra_message = ' An auto-scaling job (`--max-np` > `--num-proc`) also requires that at least two hosts ' \
+                        'are available to resolve compatible network interfaces. If you know which interfaces ' \
+                        'are compatible in your network, set `--nic` to skip this check.' if min_hosts > 1 else ''
+
+        tmout = timeout.Timeout(
+            self._start_timeout,
+            message='Timed out waiting for {{activity}}. Please check that you have '
+                    'enough resources to run at least {min_np} Horovod processes.{extra_message}'
+                    .format(min_np=min_np, extra_message=extra_message))
+
+        self._wait_hosts_cond.acquire()
+        try:
+            while True:
+                current_hosts = self._host_manager.current_hosts
+                if current_hosts.count_available_slots() >= min_np and len(current_hosts.available_hosts) >= min_hosts:
+                    return current_hosts
+                self._wait_hosts_cond.wait(tmout.remaining())
+                tmout.check_time_out_for('minimum number of hosts to become available')
+        finally:
+            self._wait_hosts_cond.release()
+
+    def _activate_hosts(self, min_np):
+        logging.info('wait for available hosts: {}'.format(min_np))
+        current_hosts = self.wait_for_available_hosts(min_np)
+        pending_slots = self._update_host_assignments(current_hosts)
+        self._worker_registry.reset(self.world_size())
+        self._start_worker_processes(pending_slots)
+
+    def _discover_hosts(self):
+        first_update = True
+        while not self._shutdown.is_set():
+            self._wait_hosts_cond.acquire()
+            try:
+                if self._host_manager.update_available_hosts():
+                    self._notify_workers_host_changes(self._host_manager.current_hosts)
+                    self._wait_hosts_cond.notify_all()
+            except RuntimeError as e:
+                if first_update:
+                    # Misconfiguration, fail the job immediately
+                    raise
+                # Transient error, retry until timeout
+                logging.warning(str(e))
+            finally:
+                self._wait_hosts_cond.release()
+            first_update = False
+            self._shutdown.wait(DISCOVER_HOSTS_FREQUENCY_SECS)
+
+    def _notify_workers_host_changes(self, current_hosts):
+        # Assignments are required to be stable via contract
+        next_host_assignments, _ = self._get_host_assignments(current_hosts)
```

Comment:
`_get_host_assignments` throws an `ValueError` when `current_hosts` is insufficient for a `min_np` assignment, which kills the discovery thread and might shutdown the whole app. An insufficient assignment is allowed from this method's point of view sometimes. The exception should be caught, ignored and make the method return. Unit test would be nice.

Reply:
Very good catch, this should have been caught by existing unit tests, but there was a bug with the way we were generating the sequence of hosts.  This should be fixed now, so we now have proper coverage of this case.

Reply:
In this case, we do still need to notify workers if the previous assignments were valid, so we should interpret it as an empty set of assignments.

Reply:
Fixed.